In [179]:
import math
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
import pandas as pd
import plotly.graph_objects as go
from pathlib import Path
import webbrowser

#### **Constants from paper**

In [223]:
# Constants (Tjek onenote)
lambda_g = 0.150608 # Guided wavelength of antenna
e_r = 4.7           # Relative dielectric of FR4
e_ff = 3.74         # Effective dielectric constant
d = 0.0241          # 2.41 cm
s = 0.0633          # 6.33 cm
L = 0.1054          # 10.54 cm
w = 0.0075          # 0.75 cm

#### **Class for defining antenna geometry**

In [ ]:
# ===============================================
# Class for Meander line coordinates
# ===============================================

class CoordinateSet:
    def __init__(self, coords, lengths):
        self.coords = np.array(coords, dtype=float)
        self.lengths = np.array(lengths, dtype=float)

    def as_array(self):
        #Return the (N,3) numpy array of (x,y,z).
        return self.coords
    
    def lengths_array(self):
        # Return the lengths
        return self.lengths
    
    def r_vectors(self):
        #Return full 3D vectors (x,y,z) as (N,3).
        return self.coords
    
    def r_norms(self):
        # Return distances from origin (N,).
        return np.linalg.norm(self.r_vectors(), axis=1)
    
    def r_hat(self):
        # Return unit vectors pointing from origin to each (x,y,z)
        r_vecs = self.r_vectors()
        return r_vecs / self.r_norms()[:,None]

# Meander line geometry [x,y,z]:
coords = [[0,-1,0], [1,0,0], [0,-1,0], [-1,0,0], [0,-1,0], [0,1,0], [0,-1,0], [-1,0,0], [-1,0,0], [1,0,0]]

# Meander line segments lengths
lengths = [0.5*d, s, d, s, d, s, d, s, d, 0.5*d]

# Create instance
meander = CoordinateSet(coords, lengths)

0.37370000000000003


Assuming a uniform microstrip with per-unit parameters as: <br>
    $Z_0 \\$
    $\gamma = \alpha + j\beta$

The current along a lossy *matched* transmission line is a single forward traveling wave: <br>
$I(\ell) = I_0 e^{-j\gamma \ell}$, <br>
where $I_0$ is the current at the input terminals and $\ell$ is the length of the transmission (meander) line

#### **Functions for the current distribution along a transmission line**

In [35]:
# ===============================================
#   Current distribution along transmission line
# ===============================================
"""
- Derived from transmission line theory
- Matched load assumed -> forward-traveling wave only (no reflections)
"""


# Lossy transmission line
def currentDistLossy(I0, l, alpha, beta):
    return I0 * np.exp(-(alpha + 1j*beta)*l)

# Lossless transmission line
def currentDistLossless(I0, l, beta):
    return I0 * np.exp(-1j*beta*l)

The E-field from an infinitesimal dipole antenna:  

$$
E_{r} = \eta \frac{I_0 l \cos(\theta)}{2\pi r^2} \left[ 1 + \frac{1}{jkr} \right] e^{-jkr}
$$

$$
E_{\theta} = j \eta \frac{k I_0 l \sin(\theta)}{4 \pi r} \left[ 1 + \frac{1}{jkr} - \frac{1}{(kr)^2} \right] e^{-jkr}
$$

$$
E_{\phi} = 0
$$

We know:

$$
\underline{\hat{r}} = \underline{\hat{s}}
$$

$$
\underline{\hat{\phi}} = \frac{\underline{\hat{h}} \times \underline{\hat{s}}}{\lVert \underline{\hat{h}} \times \underline{\hat{s}} \rVert}
$$

$$
\underline{\hat{\theta}} = \frac{(\underline{\hat{s}} \cdot \underline{\hat{h}}) \underline{\hat{s}} - \underline{\hat{h}}}{\lVert (\underline{\hat{s}} \cdot \underline{\hat{h}}) \underline{\hat{s}} - \underline{\hat{h}} \rVert}
$$

Also:

$$
R' = \lVert \underline{s} \rVert
$$

$$
\lVert \underline{\hat{h}} \times \underline{\hat{s}} \rVert = \sin \theta
$$

$$
\underline{\hat{s}} \cdot \underline{\hat{h}} = \cos \theta
$$

And:

$$
\underline{V}_{r\theta\phi} =
\begin{bmatrix}
V_r \\
V_\theta \\
V_\phi
\end{bmatrix}
=
\begin{bmatrix}
\sin\theta \cos\phi & \sin\theta \sin\phi & \cos\theta \\
\cos\theta \cos\phi & \cos\theta \sin\phi & -\sin\theta \\
-\sin\phi & \cos\phi & 0
\end{bmatrix}
\begin{bmatrix}
V_x \\
V_y \\
V_z
\end{bmatrix}
$$

where $\hat{h}$ = dipole axis unit vector, $\hat{s}$ = normalized direction vector from the Hertz dipole to the observer point.


#### **Crate dataframe for simulation**

In [318]:
# =====================================================
#   Resolution of angular space around antenna
# =====================================================

r0 = 1.0           # Fixed radius for far-field pattern
n_theta = 90       # Resolution in polar angle [0, pi/2]
n_phi   = 361      # Resolution in azimuth [0, 2*pi]

# Crate grid
theta = np.linspace(0, np.pi/2, n_theta)    # only positive z-value (0-pi/2)
phi   = np.linspace(0, 2*np.pi, n_phi)      # Full revolution (0-2pi)
TH, PH = np.meshgrid(theta, phi, indexing="ij")

# Convert spherical to Cartesian coordinates
X = r0 * np.sin(TH) * np.cos(PH)
Y = r0 * np.sin(TH) * np.sin(PH)
Z = r0 * np.cos(TH)

# Flatten into N×3 point array
points = np.stack([X.ravel(), Y.ravel(), Z.ravel()], axis=1)

# Also flatten r, theta, phi
r = np.full(points.shape[0], r0)     # constant radius
theta_flat = TH.ravel()
phi_flat   = PH.ravel()

# Create [N×2] matrix for E_r, E_t components at all points
E = np.zeros((points.shape[0], 2), dtype=complex)
H = np.zeros((points.shape[0], 1), dtype=complex)

# Build dataframe
df = pd.DataFrame({
    "X": points[:,0],
    "Y": points[:,1],
    "Z": points[:,2],
    "r": r,
    "theta": theta_flat,
    "phi": phi_flat,
    "E_r": E[:, 0],
    "E_t": E[:, 1],
    "H_p": H[:,0]
})

# Visualize dataframe
print(df.head())

     X    Y    Z    r  theta       phi       E_r       E_t       H_p
0  0.0  0.0  1.0  1.0    0.0  0.000000  0.0+0.0j  0.0+0.0j  0.0+0.0j
1  0.0  0.0  1.0  1.0    0.0  0.017453  0.0+0.0j  0.0+0.0j  0.0+0.0j
2  0.0  0.0  1.0  1.0    0.0  0.034907  0.0+0.0j  0.0+0.0j  0.0+0.0j
3  0.0  0.0  1.0  1.0    0.0  0.052360  0.0+0.0j  0.0+0.0j  0.0+0.0j
4  0.0  0.0  1.0  1.0    0.0  0.069813  0.0+0.0j  0.0+0.0j  0.0+0.0j


#### **Functions for calculating the E/H-fields**

In [ ]:
# ==========================================
#   Functions for calculating total E-field
# ==========================================

def singleDipole(I0, 
                 l_dipole, 
                 h_hat, 
                 P, 
                 freq=1060e6, 
                 free_space=True, 
                 c=None, 
                 eta=None):
    """
    Calculates the E-field from a single hertzian dipole at a single point P

    Parameters
    -----------
    I0:         Current amplitude of forward traveling wave
    l_dipole:   Length of dipole
    h_hat:      Dipole axis normal vector
    P:          Observation point of universal coordinate system [X,Y,Z]

    returns
    -----------
    Three values as radial and zenit components of E and azimuthal component of H at point P
    """

    c0 = 299_792_458    # Free space speed of light
    eta0 = 120*np.pi    # Free space impedance

    if free_space:
        c = c0
        eta = eta0
    else:
        if c is None:
            raise ValueError("Must provide some speed of light when free_space=False")
        elif eta is None:
            raise ValueError("Must provide some impedance when eta=False")

    # Constants
    Lambda = c/freq     # Wavelength
    k = 2*np.pi/Lambda  # Wave number

    # Vectors
    s_hat = P / np.linalg.norm(P)                   # s_hat      = Normal vector from dipole to obersvation point
    r = np.linalg.norm(P)                           # r          = Length from origin to observation point
    cos_th = np.dot(s_hat, h_hat)                   # cos(theta) = s_hat \dot h_hat 
    sin_th = np.linalg.norm(np.cross(h_hat, s_hat)) # sin(theta) = ||h_hat x s_hat||

    # Calculate radial E-field component
    E_r0 = eta*(I0*l_dipole*cos_th) / (2*np.pi*r**2)
    E_r1 = (1 + 1/(1j*k*r)) * np.exp(-1j*k*r)
    E_r = E_r0 * E_r1

    # Calculate zenit E-field component
    E_t0 = 1j*eta*(k*I0*l_dipole*sin_th) / (4*np.pi*r)
    E_t1 = ((1 + 1/(1j*k*r) - 1/((k*r)**2))) * np.exp(-1j*k*r)
    E_t = E_t0*E_t1

    # Calculate the azimuthal H-field component
    H_p0 = 1j*k*I0*l_dipole*sin_th / (4*np.pi*r)
    H_p1 = (1 + 1/(1j*k*r))*np.exp(-1j*k*r)
    H_p = H_p0*H_p1

    return E_r, E_t, H_p

# TEST
# E_r, E_t = singleDipole(I0=0.001, l_dipole=0.0000001, h_hat=meander.as_array()[0], P=[0,1,1]) 


def oneHertzianAllPoints(I0, 
                         l_dipole, 
                         h_hat, l_trace, 
                         dipole_idx, 
                         points, 
                         dataframe, 
                         freq=1060e6, 
                         free_space=True, 
                         c=None, 
                         eta=None, 
                         lambda_g = 0.150608):
    """
    Calculates the E- & H-field components from a single Hertzian dipole to all points in space defined in points

    Parameters
    -----------
    I0:         Current amplitude of forward traveling wave
    l_dipole:   Length of Hertzian dipole
    h_hat:      Dipole axis normal vector
    l_trace:    Meander line trace length
    dipole_idx: The index of the dipole on the trace
    points:     All points in space [X,Y,Z]
    dataframe:  Which dataframe to update

    Returns
    -----------
    Nothing - Updates the dataframe
    """

    c0 = 299_792_458    # Free space speed of light
    eta0 = 120*np.pi    # Free space impedance

    if free_space:
        c = c0
        eta = eta0
    else:
        if c is None:
            raise ValueError("Must provide some speed of light when free_space=False")
        elif eta is None:
            raise ValueError("Must provide some impedance when eta=False")

    # Constants
    Lambda = c/freq                             # Wavelength on transmission line
    beta = 2*np.pi / lambda_g                   # Phase constant of transmission line
    num_dipoles = round(l_trace / l_dipole)     # Number of dipoles on meander trace segment
    point_on_trace = num_dipoles * dipole_idx   # Which point on the trace are we? 

    # The current at dipole point on meander trace segment
    I = currentDistLossless(I0, point_on_trace, beta)

    # Iterate over all points in points and add the dipole contributions to that point in the dataframe
    for i in range(len(points)):
        E_r, E_t, H_p = singleDipole(I0, l_dipole, h_hat, points[i])
        dataframe.loc[i, "E_r"] += E_r
        dataframe.loc[i, "E_t"] += E_t
        dataframe.loc[i, "H_p"] += H_p

    return None


# TEST
#oneHertzianAllPoints(0.00001, 0.1, meander.as_array()[0], meander.lengths_array()[0], 1, points=points, dataframe=df)


def fullTraceAllPoints(I0, 
                       l_dipole, 
                       h_hat, 
                       l_trace, 
                       points, 
                       dataframe, 
                       freq=1060e6, 
                       free_space=True, 
                       c=None, 
                       eta=None, 
                       lambda_g = 0.150608):
    """
    Calculates the E- & H-field components at all points defined in points for a whole stripline

    Parameters
    -----------
    I0:         Current amplitude of forward traveling wave
    l_dipole:   Length of Hertzian dipole
    h_hat:      Dipole axis normal vector
    l_trace:    Meander line trace length
    points:     All points in space [X,Y,Z]
    dataframe:  Which dataframe to update

    Returns
    -----------
    Nothing - Updates the dataframe
    """

    # Number of dipoles on meander trace segment
    num_dipoles = round(l_trace / l_dipole)

    for idx in range(num_dipoles):
        oneHertzianAllPoints(I0, l_dipole, h_hat, l_trace, idx, points, dataframe)
        print("Calculated for dipole nr.:", idx)
    
    return None


# TEST
# fullTraceAllPoints(0.0001, 0.001, meander.as_array()[0], meander.lengths_array()[0], points=points, dataframe=df)


def meanderTotalRadiationE(I0, l_dipole, antenna, points, dataframe):
    """
    Calculates the E- & H-field components at all points defined in points for a whole Meander line

    Parameters
    -----------
    I0:         Current amplitude of forward traveling wave
    l_dipole:   Length of Hertzian dipole
    h_hat:      Dipole axis normal vector
    l_trace:    Meander line trace length
    points:     All points in space [X,Y,Z]
    dataframe:  Which dataframe to update

    Returns
    -----------
    Nothing - Updates the dataframe
    """

    num_dipoles_tot = round(sum(antenna.lengths_array()) / l_dipole)
    print("Total number of dipoles to calculate:", num_dipoles_tot)

    # Iterate fullTraceAllPoints over all traces in meander line geometry
    for trace in range(len(antenna.as_array())):
        fullTraceAllPoints(I0, l_dipole, antenna.as_array()[trace], meander.lengths_array()[trace], points, dataframe)
        
        print("Done calculating E&H-field from trace:", antenna.as_array()[trace])
    
    return None

#### **Run simulation**

In [ ]:
# Infinitesimal dipole approximation valid for l_dipole <= lambda/50
lambda0 = 3e8/1060e6
l_dipole = lambda0 / 50

# Run simulations
meanderTotalRadiationE(I0=0.001, l_dipole=0.001, antenna=meander, points=points, dataframe=df)

# Save to dataframe
df.to_csv("Meander_EH-field.csv", index=False)

Total number of dipoles to calculate: 374
Calculated for dipole nr.: 0
Calculated for dipole nr.: 1
Calculated for dipole nr.: 2
Calculated for dipole nr.: 3
Calculated for dipole nr.: 4
Calculated for dipole nr.: 5
Calculated for dipole nr.: 6


#### **Visualization of radiation pattern**
##### OBS: Denne del af koden åbner en HTML fil i din browser!

In [304]:
# ---------- 2) Compute far-field magnitude |E| ----------
Er = df["E_r"].to_numpy()
Et = df["E_t"].to_numpy()
E_mag = np.sqrt(np.abs(Er)**2 + np.abs(Et)**2)

# ---------- 3) Reshape to spherical grid ----------
theta_vals = df["theta"].to_numpy()
phi_vals   = df["phi"].to_numpy()
thetas = np.unique(theta_vals)
phis   = np.unique(phi_vals)
n_theta, n_phi = len(thetas), len(phis)

TH = theta_vals.reshape(n_theta, n_phi)
PH = phi_vals.reshape(n_theta, n_phi)
E_mag_grid = E_mag.reshape(n_theta, n_phi)

# ---------- 4) Normalize and build deformed sphere ----------
E_max = E_mag_grid.max() if E_mag_grid.size else 1.0
E_norm = (E_mag_grid / E_max) if E_max > 0 else np.zeros_like(E_mag_grid)

r0 = 1.0
R = r0 * (0.15 + 0.85 * E_norm)   # baseline prevents nulls at origin

X = R * np.sin(TH) * np.cos(PH)
Y = R * np.sin(TH) * np.sin(PH)
Z = R * np.cos(TH)

# ---------- 5) Plotly interactive surface (red = strongest) ----------
fig = go.Figure(
    data=[go.Surface(
        x=X, y=Y, z=Z,
        surfacecolor=E_norm,
        colorscale="Jet",   # high values -> red
        cmin=0, cmax=1,
        colorbar=dict(title="|E| / max")
    )]
)

fig.update_layout(
    title="3D Radiation Pattern: |E| (drag to rotate, scroll to zoom)",
    scene=dict(
        xaxis_title="X",
        yaxis_title="Y",
        zaxis_title="Z",
        aspectmode="data"
    ),
    margin=dict(l=0, r=0, t=40, b=0)
)

# ---------- 6) Save as HTML and open in default browser ----------
html_path = Path.cwd() / "radiation_pattern_3D.html"
fig.write_html(html_path, include_plotlyjs="cdn", full_html=True)
webbrowser.open(html_path.resolve().as_uri(), new=2)

True

#### **Calculate -3dB angles and directivity using simulation data**

In [303]:
# theta = 0 cut
cut = df[np.isclose(df["phi"].apply(np.real), 0.0, atol=1e-9)].copy()
cut["E_mag"] = np.sqrt(np.abs(cut["E_r"])**2 + np.abs(cut["E_t"])**2)

# Peak and half-power
E_peak = cut["E_mag"].max()
E_half = E_peak / np.sqrt(2)

# Row closest to E_half
row_half = cut.loc[(cut["E_mag"] - E_half).abs().idxmin()]

# Extract theta
theta_rad = row_half["theta"].real   # drop +0j
theta_deg = np.rad2deg(theta_rad)

print("E_half =", E_half)
print("theta at E_half (deg):", theta_deg)

# HPBW
HPBW_1d = 2*theta_deg

# Directivity approx
D = 41253 / (HPBW_1d**2) # Assuming symmetry

print("HPBW:", HPBW_1d, "\nDirectivity:", D, "=", 10*np.log10(D), "dBi")

E_half = 0.18370417577840115
theta at E_half (deg): 58.65168539325843
HPBW: 117.30337078651687 
Directivity: 2.998020186506363 = 4.768345527480203 dBi


#### **Calculating the directivity using the paper**

$$
d = 0.16\lambda_g \\
s = 0.41\lambda_g \\
L = 0.70\lambda_g \\
w = 0.05\lambda_g \\
\lambda_g = \frac{\lambda}{\sqrt{\epsilon_{ff}}} \\
\epsilon_{ff} = \frac{\epsilon_r+1}{2} + \frac{\epsilon_r-1}{2} \left[ 1 + 10\frac{h}{w} \right]^{-1/2}
$$

Assuming standard 4 layer PCB: 
$$
h = 1.6 \ mm \\
\epsilon_r = 4.5
$$

The paper mentions a necessary dimension correction $\Delta L$
$$
\Delta L = 0.5 w,
$$
where:
$$
w = \frac{c}{2f_r} \sqrt{\frac{2}{\epsilon_r + 1}}
$$
<br><br>


First we calculate $\epsilon_ff$ from knowing that,
$$
\lambda_g = \frac{\lambda}{\sqrt{\epsilon_{ff}}} \\
\lambda = c/f = \frac{3e6}{1060~\text{MHz}} = 28.3~\text{cm} \\
\epsilon_r = 4.5 \\
h = 1.6~mm \\
\therefore \epsilon_{ff} = 3.73,
$$

And therefore,
$$
\lambda_g = 0.1465~m = 14.65~cm
$$


The Meander line dimension before corrections are:
$$
d = 0.16\lambda_g = 0.02344~m \\
s = 0.42\lambda_g = 0.06153~m \\
L = 0.70\lambda_g = 0.10255~m \\
w = 0.05\lambda_g = 0.007325~m \\
$$


Now we can calculate $\Delta L$
$$
\Delta L = 0.5 \left( \frac{3*10^8~\text{m/s}}{2*1.06~\text{GHz}} \right) \sqrt{\frac{2}{4.5+1}} = 0.04266
$$

and the corrected values are:
$$
S_p = s+\Delta L = 0.10419~m \\
d_p = d + \Delta L = 0.0661~m \\
L_p = L + \Delta L = 0.14521~m
$$

Now we can calculate the HPBW's and the directivity. Knowing that:
$$
k_0 = \frac{\pi}{\sqrt{\epsilon_{ff}}L_p} = 11.20208
$$

then,
$$
\theta_E = 2.45~\text{rad} = 140\degree \\
\theta_H = 1.867 \text{rad} = 107\degree
$$

and therefore,
$$
D = \frac{41253}{\theta_E \theta_H} = 2.754 = 4.4~dBi
$$

there is hence a discrepance between calculations and simulations of:
$$
4.768~dBi-4.4~dBi = 0.368~dBi
$$